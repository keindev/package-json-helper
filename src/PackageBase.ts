import { BugsLocation } from './fields/Bugs';
import { DependencyMeta } from './fields/DependencyMeta';
import { Funding } from './fields/Funding';
import { Person } from './fields/Person';
import { Repository } from './fields/Repository';
import { JSONObject } from './types';
import { mappers } from './utils/mappers';
import { cast, parsers } from './utils/parsers';
import { check, URL_REGEXP, validators } from './utils/validators';

const NAME_MAX_LENGTH = 214;

export enum Type {
  Module = 'module',
  Commonjs = 'commonjs',
}

enum StringProps {
  Browser = 'browser',
  Description = 'description',
  License = 'license',
  Main = 'main',
  Types = 'types',
}

enum StringListProps {
  BundledDependencies = 'bundledDependencies',
  CPU = 'cpu',
  Files = 'files',
  Keywords = 'keywords',
  Man = 'man',
  OS = 'os',
  Workspaces = 'workspaces',
}

enum StringMapProps {
  Scripts = 'scripts',
  Config = 'config',
  Dependencies = 'dependencies',
  DevDependencies = 'devDependencies',
  OptionalDependencies = 'optionalDependencies',
  PeerDependencies = 'peerDependencies',
  Engines = 'engines',
  Directories = 'directories',
}

const rules = {
  name: [validators.hasValidLength('Name must be less than or equal to 214 characters', NAME_MAX_LENGTH)],
  version: [validators.isSemanticVersion('Version must be parseable by node-semver')],
  type: [validators.isEnum('Invalid package type', Object.values(Type))],
  homepage: [validators.isMatchesRegExp("Homepage can't contain any non-URL-safe characters", URL_REGEXP)],
  list: (name: string) => [validators.isStringArray(`${name} must be array of strings`)],
  map: (name: string) => [validators.isKeyValueObject(`${name} must be key-value object`)],
};

type IStringProps = { [key in StringProps]: string };
type IStringListProps = { [key in StringListProps]: Set<string> };
type IStringMapProps = { [key in StringMapProps]: Map<string, string> };

class PackageBase implements IStringProps, IStringListProps, IStringMapProps {
  /** If set to `true`, then npm will refuse to publish it. */
  private: boolean;
  /** Package description, listed in `npm search`. */
  description = '';
  /** The license for the package. */
  license = '';
  /** The module ID that is the primary entry point to the program. */
  main = '';
  /** Point to bundled declaration file. */
  types = '';
  /** Client-side module ID that is the primary entry point. */
  browser = '';

  #name: string;
  #version: string;
  #homepage: string;
  #type: Type;
  #keywords = new Set<string>();
  #files = new Set<string>();
  #man = new Set<string>();
  #bundledDependencies = new Set<string>();
  #os = new Set<string>();
  #cpu = new Set<string>();
  #workspaces = new Set<string>();
  #bin = new Map<string, string>();
  #scripts = new Map<string, string>();
  #config = new Map<string, string>();
  #dependencies = new Map<string, string>();
  #devDependencies = new Map<string, string>();
  #optionalDependencies = new Map<string, string>();
  #peerDependencies = new Map<string, string>();
  #engines = new Map<string, string>();
  #directories = new Map<string, string>();
  #bugs: BugsLocation;
  #author: Person;
  #repository: Repository;
  #contributors: Map<string, Person>;
  #maintainers: Map<string, Person>;
  #funding: Map<string, Funding>;
  #peerDependenciesMeta: Map<string, DependencyMeta>;
  #publishConfig: Map<string, string | number | boolean>;

  protected data: JSONObject;

  constructor(data: JSONObject) {
    this.data = data;
    this.private = !!data.private;
    this.#name = parsers.getString(rules.name)(data.name);
    this.#version = parsers.getString(rules.version)(data.version);
    this.#homepage = parsers.getString(rules.homepage)(data.homepage);
    this.#type = data.type ? (parsers.getString(rules.type)(data.type) as Type) : Type.Commonjs;
    this.#bugs = new BugsLocation(data.bugs);
    this.#author = new Person(data.author);
    this.#repository = new Repository(data.repository);
    this.#bin = mappers.toBin(data.bin);
    this.#contributors = mappers.toPersons(data.contributors);
    this.#maintainers = mappers.toPersons(data.maintainers);
    this.#funding = mappers.toFunding(data.funding);
    this.#peerDependenciesMeta = mappers.toDependencyMeta(data.peerDependenciesMeta);
    this.#publishConfig = mappers.toPublishConfig(data.publishConfig);

    Object.values(StringProps).forEach(name => (this[name] = cast.toString(data[name])));
    Object.values(StringListProps).forEach(name =>
      (parsers.getArray(rules.list(name))(data[name]) as string[]).forEach(item => this[name].add(item))
    );
    Object.values(StringMapProps).forEach(name =>
      Object.entries(parsers.getObject(rules.map(name))(data[name]) as { [key: string]: string }).forEach(
        ([key, value]) => this[name].set(key, value)
      )
    );
  }

  /** Name of the package */
  get name(): string {
    return this.#name;
  }

  set name(value: string) {
    this.#name = check(value, rules.name);
  }

  /** Package version, parseable by [`node-semver`](https://github.com/npm/node-semver). */
  get version(): string {
    return this.#version;
  }

  set version(value: string) {
    this.#version = check(value, rules.version);
  }

  /** Resolution algorithm for importing ".js" files from the package's scope. */
  get type(): Type {
    return this.#type;
  }

  set type(value: Type) {
    this.#type = check(value, rules.type);
  }

  /** The URL to the package's homepage. */
  get homepage(): string {
    return this.#homepage;
  }

  set homepage(value: string) {
    this.#homepage = check(value, rules.homepage);
  }

  /** Keywords associated with package, listed in `npm search`. */
  get keywords(): Set<string> {
    return this.#keywords;
  }

  /** The files included in the package. */
  get files(): Set<string> {
    return this.#files;
  }

  /** Filenames to put in place for the `man` program to find. */
  get man(): Set<string> {
    return this.#man;
  }

  /** Operating systems the module runs on. */
  get os(): Set<string> {
    return this.#os;
  }

  /** CPU architectures the module runs on. */
  get cpu(): Set<string> {
    return this.#cpu;
  }

  /** Array of file patterns that describes locations within the local file system that the install client should look up to find each workspace that needs to be symlinked to the top level node_modules folder */
  get workspaces(): Set<string> {
    return this.#workspaces;
  }

  /** Package names that are bundled when the package is published. */
  get bundledDependencies(): Set<string> {
    return this.#bundledDependencies;
  }

  /** The executable files that should be installed into the `PATH`. */
  get bin(): Map<string, string> {
    return this.#bin;
  }

  /** Script commands that are run at various times in the lifecycle of the package. The key is the lifecycle event, and the value is the command to run at that point. */
  get scripts(): Map<string, string> {
    return this.#scripts;
  }

  /** Is used to set configuration parameters used in package scripts that persist across upgrades. */
  get config(): Map<string, string> {
    return this.#config;
  }

  /** The dependencies of the package. */
  get dependencies(): Map<string, string> {
    return this.#dependencies;
  }

  /** Additional tooling dependencies that are not required for the package to work. Usually test, build, or documentation tooling. */
  get devDependencies(): Map<string, string> {
    return this.#devDependencies;
  }

  /** Dependencies that are skipped if they fail to install. */
  get optionalDependencies(): Map<string, string> {
    return this.#optionalDependencies;
  }

  /** Dependencies that will usually be required by the package user directly or via another dependency. */
  get peerDependencies(): Map<string, string> {
    return this.#peerDependencies;
  }

  /** Engines that this package runs on. */
  get engines(): Map<string, string> {
    return this.#engines;
  }

  /** Indicates the structure of the package. */
  get directories(): Map<string, string> {
    return this.#directories;
  }

  /** The URL to the package's issue tracker and/or the email address to which issues should be reported. */
  get bugs(): BugsLocation {
    return this.#bugs;
  }

  /** Author of the package */
  get author(): Person {
    return this.#author;
  }

  /** Location for the code repository. */
  get repository(): Repository {
    return this.#repository;
  }

  /** A list of people who contributed to the package. */
  get contributors(): Map<string, Person> {
    return this.#contributors;
  }

  /** A list of people who maintain the package. */
  get maintainers(): Map<string, Person> {
    return this.#maintainers;
  }

  /** Describes and notifies consumers of a package's monetary support information. */
  get funding(): Map<string, Funding> {
    return this.#funding;
  }

  /** Indicate peer dependencies that are optional. */
  get peerDependenciesMeta(): Map<string, DependencyMeta> {
    return this.#peerDependenciesMeta;
  }

  /** A set of config values that will be used at publish-time. It's especially handy to set the tag, registry or access, to ensure that a given package is not tagged with 'latest', published to the global public registry or that a scoped module is private by default. */
  get publishConfig(): Map<string, string | number | boolean> {
    return this.#publishConfig;
  }
}

export default PackageBase;
