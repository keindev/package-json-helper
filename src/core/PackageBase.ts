import { BugsLocation } from '../fields/Bugs';
import { Dependency } from '../fields/Dependency';
import { DependencyMeta } from '../fields/DependencyMeta';
import { ExportMap } from '../fields/ExportMap';
import Field from '../fields/Field';
import { Funding } from '../fields/Funding';
import { ImportMap } from '../fields/ImportMap';
import { Person } from '../fields/Person';
import { Repository } from '../fields/Repository';
import {
    DependenciesMapPropsMap, IPackageProps, JSONObject, JSONValue, StringListPropsMap, StringMapPropsMap,
    StringPropsMap,
} from '../types';
import { cast } from '../utils/parsers';
import AbstractPackage from './AbstractPackage';

const TAB_WIDTH = 2;

class PackageBase extends AbstractPackage implements IPackageProps {
  /** Author of the package */
  author?: Person;
  /** The executable files that should be installed into the `PATH`. */
  readonly bin: Map<string, string>;
  /** Client-side module ID that is the primary entry point. */
  browser?: string;
  /** The URL to the package's issue tracker and/or the email address to which issues should be reported. */
  bugs?: BugsLocation;
  /** Package names that are bundled when the package is published. */
  readonly bundledDependencies = new Set<string>();
  /** Is used to set configuration parameters used in package scripts that persist across upgrades. */
  readonly config = new Map<string, string>();
  /** A list of people who contributed to the package. */
  readonly contributors: Map<string, Person>;
  /** CPU architectures the module runs on. */
  readonly cpu = new Set<string>();
  /** The dependencies of the package. */
  readonly dependencies = new Map<string, Dependency>();
  /** Package description, listed in `npm search`. */
  description?: string;
  /** Additional tooling dependencies that are not required for the package to work. Usually test, build, or documentation tooling. */
  readonly devDependencies = new Map<string, Dependency>();
  /** Indicates the structure of the package. */
  readonly directories = new Map<string, string>();
  /** Engines that this package runs on. */
  readonly engines = new Map<string, string>();
  /** Standard entry points of the package, with enhanced support for ECMAScript Modules. */
  readonly exports: ExportMap;
  /** The files included in the package. */
  readonly files = new Set<string>();
  /** Describes and notifies consumers of a package's monetary support information. */
  readonly funding: Map<string, Funding>;
  /** Import maps that only apply to import specifiers from within the package itself */
  readonly imports: ImportMap;
  /** Keywords associated with package, listed in `npm search`. */
  readonly keywords = new Set<string>();
  /** The license for the package. */
  license?: string;
  /** The module ID that is the primary entry point to the program. */
  main?: string;
  /** A list of people who maintain the package. */
  readonly maintainers: Map<string, Person>;
  /** Filenames to put in place for the `man` program to find. */
  readonly man = new Set<string>();
  /** Dependencies that are skipped if they fail to install. */
  readonly optionalDependencies = new Map<string, Dependency>();
  /** Operating systems the module runs on. */
  readonly os = new Set<string>();
  /** Dependencies that will usually be required by the package user directly or via another dependency. */
  readonly peerDependencies = new Map<string, Dependency>();
  /** Indicate peer dependencies that are optional. */
  readonly peerDependenciesMeta: Map<string, DependencyMeta>;
  /** If set to `true`, then npm will refuse to publish it. */
  private: boolean;
  /** A set of config values that will be used at publish-time. It's especially handy to set the tag, registry or access, to ensure that a given package is not tagged with 'latest', published to the global public registry or that a scoped module is private by default. */
  readonly publishConfig: Map<string, string | number | boolean>;
  /** Location for the code repository. */
  repository?: Repository;
  /** Script commands that are run at various times in the lifecycle of the package. The key is the lifecycle event, and the value is the command to run at that point. */
  readonly scripts = new Map<string, string>();
  /** Point to bundled declaration file. */
  types?: string;
  /** Array of file patterns that describes locations within the local file system that the install client should look up to find each workspace that needs to be symlinked to the top level node_modules folder */
  readonly workspaces = new Set<string>();

  constructor(data: JSONObject) {
    super(data);

    this.private = !!data.private;
    this.bugs = cast.toBugsLocation(data.bugs);
    this.author = cast.toPerson(data.author);
    this.repository = cast.toRepository(data.repository);
    this.bin = cast.toBin(this.nameWithoutScope, data.bin);
    this.contributors = cast.toPersons(data.contributors);
    this.maintainers = cast.toPersons(data.maintainers);
    this.funding = cast.toFundingList(data.funding);
    this.peerDependenciesMeta = cast.toDependencyMeta(data.peerDependenciesMeta);
    this.publishConfig = cast.toPublishConfig(data.publishConfig);
    this.exports = cast.toExportsMap(data.exports);
    this.imports = cast.toImportsMap(data.imports);

    StringPropsMap.forEach(name => (this[name] = cast.toString(data[name])));
    StringListPropsMap.forEach(name => cast.toSet(data[name], this[name]));
    StringMapPropsMap.forEach(name => cast.toStringMap(data[name], this[name]));
    DependenciesMapPropsMap.forEach(name => cast.toDependencyMap(data[name], this[name]));
  }

  // eslint-disable-next-line max-lines-per-function
  toString(): string {
    const OBJECT_FIELDS_NAMES = ['contributors', 'maintainers', 'funding'];

    return JSON.stringify(
      {
        name: this.name,
        version: this.version,
        description: this.description,
        license: this.license,
        private: this.private || undefined,
        homepage: this.homepage,
        author: this.author,
        contributors: this.contributors,
        maintainers: this.maintainers,
        keywords: this.keywords,
        funding: this.funding,
        bugs: this.bugs,
        types: this.types,
        browser: this.browser,
        type: this.type,
        files: this.files,
        workspaces: this.workspaces,
        directories: this.directories,
        main: this.exports.map.size ? undefined : this.main,
        imports: this.imports,
        exports: this.exports,
        bin: this.bin,
        man: this.man,
        repository: this.repository,
        config: this.config,
        publishConfig: this.publishConfig,
        os: this.os,
        cpu: this.cpu,
        engines: this.engines,
        scripts: this.scripts,
        dependencies: this.dependencies,
        devDependencies: this.devDependencies,
        peerDependencies: this.peerDependencies,
        peerDependenciesMeta: this.peerDependenciesMeta,
        bundledDependencies: this.bundledDependencies,
        optionalDependencies: this.optionalDependencies,
      },
      (key: string, value: JSONValue): JSONValue => {
        if (value === null || ((typeof value === 'string' || Array.isArray(value)) && !value.length)) return undefined;
        if (value instanceof Set) return value.size ? [...value.values()].sort() : undefined;
        if (value instanceof Map) {
          if (!value.size) return undefined;
          if (OBJECT_FIELDS_NAMES.some(field => field === key)) return [...value.values()];

          return Object.fromEntries([...value.entries()].sort(([a], [b]) => a.localeCompare(b)));
        }
        if (typeof value === 'object') {
          if (Object.keys(value).length === 0) return undefined;
          if (value instanceof Field) return value.getSnapshot();
        }

        return value;
      },
      TAB_WIDTH
    );
  }
}

export default PackageBase;
