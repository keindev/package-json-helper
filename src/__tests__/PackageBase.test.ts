import fs from 'fs';
import path from 'path';

import PackageBase from '../core/PackageBase';

const base = {
  ...JSON.parse(fs.readFileSync(path.resolve(process.cwd(), 'package.json'), 'utf-8')),
};

describe('PackageBase', () => {
  it('Parse base', () => {
    const pkg = new PackageBase({ ...base, bin: './index.js' });

    expect(pkg.name).toBe(base.name);
    expect(pkg.author?.name).toBe('Daniil Ryazanov');
    expect(pkg.author?.email).toBe('kein@tagproject.ru');
    expect(pkg.toString()).toMatchSnapshot();
  });

  it('Parse shorts', () => {
    const pkg = new PackageBase({
      ...base,
      author: 'Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)',
      bugs: { url: 'https://github.com/owner/project/issues', email: 'project@hostname.com' },
      funding: [
        { type: 'individual', url: 'http://example.com/donate' },
        'http://example.com/donateAlso',
        { type: 'patreon', url: 'https://www.patreon.com/my-account' },
      ],
      bin: { 'my-program': './path/to/program' },
      man: ['./man/foo.1', './man/bar.1'],
      repository: 'github:user/repo',
      dependencies: { foo: 'user/repo#feature/branch', bar: 'file:../foo/bar' },
      peerDependencies: { tea: '2.x', 'soy-milk': '1.2' },
      peerDependenciesMeta: { 'soy-milk': { optional: true } },
      bundledDependencies: ['super-streams'],
      os: ['darwin', 'linux'],
      engines: { npm: '~1.0.20' },
      cpu: ['!arm', '!mips'],
      workspaces: ['./packages/*'],
    });

    expect(pkg.author?.name).toBe('Barney Rubble');
    expect(pkg.author?.email).toBe('b@rubble.com');
    expect(pkg.author?.url).toBe('http://barnyrubble.tumblr.com/');
    expect(pkg.bugs?.url).toBe('https://github.com/owner/project/issues');
    expect(pkg.funding.size).toBe(3);
    expect(pkg.funding.has('http://example.com/donate')).toBeTruthy();
    expect(pkg.bin.get('my-program')).toBe('./path/to/program');
    expect(pkg.man?.size).toBe(2);
    expect(pkg.os?.size).toBe(2);
    expect(pkg.cpu?.size).toBe(2);
    expect(pkg.engines?.size).toBe(1);
    expect(pkg.dependencies?.size).toBe(2);
    expect(pkg.peerDependencies?.size).toBe(2);
    expect(pkg.peerDependenciesMeta.get('soy-milk')?.optional).toBeTruthy();
    expect(pkg.bundledDependencies?.size).toBe(1);
    expect(pkg.workspaces?.size).toBe(1);
    expect(pkg.toString()).toMatchSnapshot();

    console.log(pkg.toString());
  });
});
