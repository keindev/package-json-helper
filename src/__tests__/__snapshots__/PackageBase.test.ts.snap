// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageBase Parse base 1`] = `
"{
  \\"name\\": \\"package-json-helper\\",
  \\"version\\": \\"0.9.0\\",
  \\"description\\": \\"Helper for updating, comparing and fetching changes in package.json\\",
  \\"license\\": \\"MIT\\",
  \\"homepage\\": \\"https://github.com/keindev/package-json-helper#readme\\",
  \\"author\\": {
    \\"name\\": \\"Daniil Ryazanov\\",
    \\"email\\": \\"kein@tagproject.ru\\"
  },
  \\"keywords\\": [
    \\"package\\",
    \\"json\\",
    \\"helper\\",
    \\"util\\"
  ],
  \\"types\\": \\"lib/index.d.ts\\",
  \\"type\\": \\"module\\",
  \\"main\\": \\"lib/index.js\\",
  \\"bin\\": {
    \\"\\": \\"./index.js\\"
  },
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"git+https://github.com/keindev/package-json-helper.git\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=14.0.0\\"
  },
  \\"scripts\\": {
    \\"build\\": \\"npm-run-all build:*\\",
    \\"build:ts\\": \\"rimraf lib && tsc --extendedDiagnostics\\",
    \\"generate\\": \\"npm-run-all generate:*\\",
    \\"generate:changelog\\": \\"node --experimental-specifier-resolution=node node_modules/.bin/changelog generate --bump\\",
    \\"generate:ghinfo\\": \\"node --experimental-specifier-resolution=node node_modules/.bin/ghinfo generate -d media -t utils\\",
    \\"lint\\": \\"npm-run-all lint:*\\",
    \\"lint:eslint\\": \\"rimraf coverage && eslint \\\\\\"src/**\\\\\\" --ext .ts\\",
    \\"lint:spell\\": \\"cspell -c .vscode/cspell.json --no-summary \\\\\\"README.md\\\\\\" \\\\\\"src/**/*.ts\\\\\\"\\",
    \\"prepare\\": \\"npm-run-all prepare:*\\",
    \\"prepare:config\\": \\"node --experimental-specifier-resolution=node node_modules/.bin/ts-package-shared-config\\",
    \\"prepare:husky\\": \\"husky install\\",
    \\"release\\": \\"npm-run-all prepare lint test build generate\\",
    \\"test\\": \\"npm-run-all test:*\\",
    \\"test:jest\\": \\"node --experimental-vm-modules node_modules/.bin/jest\\"
  },
  \\"dependencies\\": {
    \\"lodash.clonedeep\\": \\"^4.5.0\\",
    \\"semver\\": \\"^7.3.5\\"
  },
  \\"devDependencies\\": {
    \\"@tagproject/ts-package-shared-config\\": \\"^4.0.0\\",
    \\"@types/jest\\": \\"^27.0.1\\",
    \\"@types/lodash.clonedeep\\": \\"^4.5.6\\",
    \\"@types/node\\": \\"^16.7.0\\",
    \\"@types/semver\\": \\"^7.3.8\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^4.29.2\\",
    \\"@typescript-eslint/parser\\": \\"^4.29.2\\",
    \\"changelog-guru\\": \\"^3.0.2\\",
    \\"cspell\\": \\"^5.7.2\\",
    \\"eslint\\": \\"^7.32.0\\",
    \\"eslint-config-prettier\\": \\"^8.3.0\\",
    \\"eslint-plugin-import\\": \\"^2.24.1\\",
    \\"eslint-plugin-jest\\": \\"^24.4.0\\",
    \\"eslint-plugin-node\\": \\"^11.1.0\\",
    \\"eslint-plugin-optimize-regex\\": \\"^1.2.1\\",
    \\"eslint-plugin-promise\\": \\"^5.1.0\\",
    \\"ghinfo\\": \\"^2.0.3\\",
    \\"husky\\": \\"^7.0.1\\",
    \\"jest\\": \\"^27.0.6\\",
    \\"npm-run-all\\": \\"^4.1.5\\",
    \\"prettier\\": \\"^2.3.2\\",
    \\"rimraf\\": \\"^3.0.2\\",
    \\"ts-jest\\": \\"^27.0.5\\",
    \\"typescript\\": \\"^4.3.5\\"
  }
}"
`;

exports[`PackageBase Parse shorts 1`] = `
"{
  \\"name\\": \\"package-json-helper\\",
  \\"version\\": \\"0.9.0\\",
  \\"description\\": \\"Helper for updating, comparing and fetching changes in package.json\\",
  \\"license\\": \\"MIT\\",
  \\"homepage\\": \\"https://github.com/keindev/package-json-helper#readme\\",
  \\"author\\": \\"Barney Rubble <b@rubble.com> (http://barnyrubble.tumblr.com/)\\",
  \\"keywords\\": [
    \\"package\\",
    \\"json\\",
    \\"helper\\",
    \\"util\\"
  ],
  \\"funding\\": {
    \\"http://example.com/donate\\": {
      \\"url\\": \\"http://example.com/donate\\",
      \\"type\\": \\"individual\\"
    },
    \\"http://example.com/donateAlso\\": \\"http://example.com/donateAlso\\",
    \\"https://www.patreon.com/my-account\\": {
      \\"url\\": \\"https://www.patreon.com/my-account\\",
      \\"type\\": \\"patreon\\"
    }
  },
  \\"types\\": \\"lib/index.d.ts\\",
  \\"type\\": \\"module\\",
  \\"workspaces\\": [
    \\"./packages/*\\"
  ],
  \\"main\\": \\"lib/index.js\\",
  \\"bin\\": {
    \\"my-program\\": \\"./path/to/program\\"
  },
  \\"man\\": [
    \\"./man/foo.1\\",
    \\"./man/bar.1\\"
  ],
  \\"repository\\": \\"github:user/repo\\",
  \\"os\\": [
    \\"darwin\\",
    \\"linux\\"
  ],
  \\"cpu\\": [
    \\"!arm\\",
    \\"!mips\\"
  ],
  \\"engines\\": {
    \\"npm\\": \\"~1.0.20\\"
  },
  \\"scripts\\": {
    \\"build\\": \\"npm-run-all build:*\\",
    \\"build:ts\\": \\"rimraf lib && tsc --extendedDiagnostics\\",
    \\"generate\\": \\"npm-run-all generate:*\\",
    \\"generate:changelog\\": \\"node --experimental-specifier-resolution=node node_modules/.bin/changelog generate --bump\\",
    \\"generate:ghinfo\\": \\"node --experimental-specifier-resolution=node node_modules/.bin/ghinfo generate -d media -t utils\\",
    \\"lint\\": \\"npm-run-all lint:*\\",
    \\"lint:eslint\\": \\"rimraf coverage && eslint \\\\\\"src/**\\\\\\" --ext .ts\\",
    \\"lint:spell\\": \\"cspell -c .vscode/cspell.json --no-summary \\\\\\"README.md\\\\\\" \\\\\\"src/**/*.ts\\\\\\"\\",
    \\"prepare\\": \\"npm-run-all prepare:*\\",
    \\"prepare:config\\": \\"node --experimental-specifier-resolution=node node_modules/.bin/ts-package-shared-config\\",
    \\"prepare:husky\\": \\"husky install\\",
    \\"release\\": \\"npm-run-all prepare lint test build generate\\",
    \\"test\\": \\"npm-run-all test:*\\",
    \\"test:jest\\": \\"node --experimental-vm-modules node_modules/.bin/jest\\"
  },
  \\"dependencies\\": {
    \\"foo\\": \\"user/repo#feature/branch\\",
    \\"bar\\": \\"file:../foo/bar\\"
  },
  \\"devDependencies\\": {
    \\"@tagproject/ts-package-shared-config\\": \\"^4.0.0\\",
    \\"@types/jest\\": \\"^27.0.1\\",
    \\"@types/lodash.clonedeep\\": \\"^4.5.6\\",
    \\"@types/node\\": \\"^16.7.0\\",
    \\"@types/semver\\": \\"^7.3.8\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"^4.29.2\\",
    \\"@typescript-eslint/parser\\": \\"^4.29.2\\",
    \\"changelog-guru\\": \\"^3.0.2\\",
    \\"cspell\\": \\"^5.7.2\\",
    \\"eslint\\": \\"^7.32.0\\",
    \\"eslint-config-prettier\\": \\"^8.3.0\\",
    \\"eslint-plugin-import\\": \\"^2.24.1\\",
    \\"eslint-plugin-jest\\": \\"^24.4.0\\",
    \\"eslint-plugin-node\\": \\"^11.1.0\\",
    \\"eslint-plugin-optimize-regex\\": \\"^1.2.1\\",
    \\"eslint-plugin-promise\\": \\"^5.1.0\\",
    \\"ghinfo\\": \\"^2.0.3\\",
    \\"husky\\": \\"^7.0.1\\",
    \\"jest\\": \\"^27.0.6\\",
    \\"npm-run-all\\": \\"^4.1.5\\",
    \\"prettier\\": \\"^2.3.2\\",
    \\"rimraf\\": \\"^3.0.2\\",
    \\"ts-jest\\": \\"^27.0.5\\",
    \\"typescript\\": \\"^4.3.5\\"
  },
  \\"peerDependencies\\": {
    \\"tea\\": \\"2.x\\",
    \\"soy-milk\\": \\"1.2\\"
  },
  \\"peerDependenciesMeta\\": {
    \\"soy-milk\\": {
      \\"optional\\": true
    }
  },
  \\"bundledDependencies\\": [
    \\"super-streams\\"
  ]
}"
`;
